service/controller/appcatalog/resource/appcatalogentry/create.go

```diff
@@ -12,5 +24,165 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+			}
+
+			created++
+		}
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created %d updated %d appcatalogentries for catalog %#q", created, updated, cr.Name))
+
+	return nil
+}
+
+func (r *Resource) createAppCatalogEntry(ctx context.Context, entryCR *v1alpha1.AppCatalogEntry) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating appcatalogentry CR %#q in namespace %#q", entryCR.Name, entryCR.Namespace))
+
+	_, err := r.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogEntries(entryCR.Namespace).Create(ctx, entryCR, metav1.CreateOptions{})
+	if apierrors.IsAlreadyExists(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("already created appcatalogentry CR %#q in namespace %#q", entryCR.Name, entryCR.Namespace))
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created appcatalogentry CR %#q in namespace %#q", entryCR.Name, entryCR.Namespace))
+
+	return nil
+}
+
+func (r *Resource) updateAppCatalogEntry(ctx context.Context, entryCR *v1alpha1.AppCatalogEntry) error {
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating appcatalogentry CR %#q in namespace %#q", entryCR.Name, entryCR.Namespace))
+
+	currentCR, err := r.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogEntries(entryCR.Namespace).Get(ctx, entryCR.Name, metav1.GetOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	entryCR.ResourceVersion = currentCR.ResourceVersion
+	_, err = r.k8sClient.G8sClient().ApplicationV1alpha1().AppCatalogEntries(entryCR.Namespace).Update(ctx, entryCR, metav1.UpdateOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated appcatalogentry CR %#q in namespace %#q", entryCR.Name, entryCR.Namespace))
+
 	return nil
 }
+
+func newAppCatalogEntries(ctx context.Context, cr v1alpha1.AppCatalog, index index) (map[string]*v1alpha1.AppCatalogEntry, error) {
+	entryCRs := map[string]*v1alpha1.AppCatalogEntry{}
+
+	for _, entries := range index.Entries {
+		for _, entry := range entries {
+			name := fmt.Sprintf("%s-%s-%s", cr.Name, entry.Name, entry.Version)
+
+			createdTime, err := parseTime(entry.Created)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+
+			// Until we add support for metadata files the updated time will be
+			// the same as the created time.
+			updatedTime := createdTime
+
+			entryCR := &v1alpha1.AppCatalogEntry{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: apiVersion,
+					Kind:       kindAppCatalogEntry,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: metav1.NamespaceDefault,
+					Labels: map[string]string{
+						pkglabel.AppKubernetesName: entry.Name,
+						pkglabel.CatalogName:       cr.Name,
+						pkglabel.CatalogType:       key.CatalogType(cr),
+						label.ManagedBy:            project.Name(),
```

Comment:
I know that it's only app-operator-unique would do this job, but can we put the additional label to indicate that fact clearly?

Reply:
Yes good point. I'd rather not add another label but I created a `key.AppCatalogEntryManagedBy` func that returns app-operator unique.